duckdb-vortex/src/vortex_extension.cpp

```diff
@@ -49,51 +50,82 @@ struct VortexBindData : public TableFunctionData {
 /// Local state for the Vortex table function that tracks the progress of a scan
 /// operation. In DuckDB's execution model, a query reading from a file can be
 /// parallelized by dividing it into ranges, each handled by a different scan.
-struct VortexScanState : public LocalTableFunctionState {
+struct VortexScanLocalState : public LocalTableFunctionState {
 	idx_t current_row = 0;
 	bool finished = false;
 	mutable Array *array;
+};
+
+struct VortexScanGlobalState : public GlobalTableFunctionState {
+	ArrayStream *array_stream;
+	std::mutex stream_lock;
+	bool finished;
 
 	optional_ptr<TableFilterSet> filter;
 	// The column idx that must be returned by the scan.
 	vector<idx_t> column_ids;
-};
 
-struct VortexScanSharedState : public GlobalTableFunctionState {
+	vector<idx_t> projection_ids;
+
 	idx_t MaxThreads() const override {
-		return 1;
+		return 99999;
 	}
 };
 
 static void VortexScanFunction(ClientContext &context, TableFunctionInput &data, DataChunk &output) {
-	auto &bind_data = data.bind_data->Cast<VortexBindData>(); // NOLINT
-	auto &state = data.local_state->Cast<VortexScanState>();  // NOLINT
-
-	if (bind_data.array_stream == nullptr) {
-		const char **const column_names = new const char *[state.column_ids.size()];
-		auto idx = 0;
-		for (auto col_id : state.column_ids) {
-			column_names[idx++] = bind_data.column_names[col_id].c_str();
-		}
-		auto options = FileScanOptions {column_names, static_cast<int>(state.column_ids.size())};
-
-		bind_data.array_stream = File_scan(bind_data.file, &options);
-
-		delete[] column_names;
-	}
+	auto &bind_data = data.bind_data->Cast<VortexBindData>();         // NOLINT
+	auto &state = data.local_state->Cast<VortexScanLocalState>();     // NOLINT
+	auto &g_state = data.global_state->Cast<VortexScanGlobalState>(); // NOLINT
 
 	if (state.finished) {
 		return;
 	}
 
 	if (state.array == nullptr) {
-		auto next = FFIArrayStream_next(bind_data.array_stream);
+		std::lock_guard l(g_state.stream_lock);
+
+		if (g_state.finished) {
+			return;
+		}
+
+		if (g_state.array_stream == nullptr) {
+			auto column_names = std::vector<char *>();
+			for (auto col_id : g_state.projection_ids) {
+				column_names.push_back(const_cast<char *>(bind_data.column_names[col_id].c_str()));
+			}
+
+			auto str = std::string("");
+
+			if (g_state.filter != nullptr) {
+				google::protobuf::Arena arena;
+				vector<vortex::expr::Expr *> exprs;
+				for (const auto &[col_id, value] : g_state.filter->filters) {
+					auto col_name = bind_data.column_names[g_state.column_ids[col_id]];
+					auto conj = table_expression_into_expr(arena, *value, col_name);
+					exprs.push_back(conj);
+				}
+				auto expr = flatten_exprs(arena, exprs);
+				str = expr->SerializeAsString();
+			}
+
+			auto options = FileScanOptions {
+			    .projection = column_names.data(),
+			    .projection_len = static_cast<int>(g_state.projection_ids.size()),
```

Comment:
Curious that `projection_len` is an int.

Reply:
Ye, likely an oversight on the FFI impl side

Reply:
I guess that langs without unsigned values care about this